text/0000-trinary-trytes/0000-trinary-trytes.md

```diff
@@ -0,0 +1,117 @@
++ Feature Name: `trinary-trytes`
++ Start Date: 2019-09-09
++ RFC PR: [iotaledger/bee-rfcs#4](https://github.com/iotaledger/bee-rfcs/pull/4)
++ Bee issue: [iotaledger/bee#50](https://github.com/iotaledger/bee/issues/50)
+
+# Summary
+
+Define type `Trytes` for higher level usage. `Trytes` is going to be similar to
+`String` which is a vector of `u8` under the structure but with the coverage of
+`BytesMut` from tokio's bytes crate. Bytes from tokio has many zero-copy
+features we can utilize, So we can initialize a `Trytes` from a slice without
+heap allocation. `Trytes` provides almost same but restricted methods with
+different signatures compares to `String`, because we are going to make sure
+every inputs to `Trytes` are valid **tryte alphabete**. **Tryte alphabete** is a
+table to make trytes easy to read. Since it serves this purpose and would
+interact with underlying ternary operations in most cases, `Trytes` will behave
+like string literal and each charaters are actually UTF-8 encoded character. So
+`9` has value of 57 and `A` to `Z` have values of `65` to `90` in its vector.
+
+# Motivation
+
+The reason to implement trytes only at the moment is not only to be loosly
+coupled from **trits**, but also layout what a ternary type could looks like. We
+will eventually have various encoding techniques. It may become a mess if we
+implement all in one structure, or seperate files but without a well-defined
+layout. For the reason to be `BytesMut` is because we want to take advantage of
+its zero-copy characteristic. We can also transfer into the `Buf` or `BufMut`
+defined by `bytes` for the later usage of `trit` encoding. I also would like to
+make it be able to represent as string literals and slice of values at the same
+time. As for choice of tryte alphabete being UTF-8 encoded is because we expect
+most users would like to conver strings and characters to `Trytes`. It makes
+more scene to stay same encoding as others in Rust. Maybe we will need to
+convert byte literals in the end, but still this can be done afterwards.
+
+# Detailed design
+
+## Trytes
+
+Type `Trytes` is a type `BytesMut` and moreover a smart pointer. There are also
+a series of fixed sized type to help checking at compiled time to increase
+correctness and performance.
+The structure simply looks like this:
+
+```rust
+pub type Tryte = u8;
+
+pub type Tag = [Tryte; 9];
+pub type Adress = [Tryte; 81];
```

Comment:
I don't think that these typedef should be in the Trinary crate. They should be in the model crate. What we could do here is having `pub type Tryte9 = [Tryte; 9];` and later define `Tag`.

Reply:
Sure, the definition can change. I just give some examples. Will update to this later.